stdlib/public/Concurrency/Task.cpp

```diff
@@ -534,32 +561,93 @@ static AsyncTaskAndContext swift_task_create_commonImpl(
   }
 
   AsyncTask *parent = nullptr;
+  AsyncTask *currentTask = swift_task_getCurrent();
   if (jobFlags.task_isChildTask()) {
-    parent = swift_task_getCurrent();
+    parent = currentTask;
     assert(parent != nullptr && "creating a child task with no active task");
   }
 
-  // Inherit the priority of the currently-executing task if unspecified and
-  // we want to inherit.
-  if (jobFlags.getPriority() == JobPriority::Unspecified &&
-      (jobFlags.task_isChildTask() || taskCreateFlags.inheritContext())) {
-    AsyncTask *currentTask = parent;
-    if (!currentTask)
-      currentTask = swift_task_getCurrent();
-
-    if (currentTask)
-      jobFlags.setPriority(currentTask->getPriority());
-    else if (taskCreateFlags.inheritContext())
-      jobFlags.setPriority(swift_task_getCurrentThreadPriority());
+  // Start with user specified priority at creation time (if any)
+  JobPriority basePriority = (taskCreateFlags.getCreationPriority());
+
+
+  if (swift_task_is_detached(taskCreateFlags, jobFlags)) {
+    // Case 1: No priority specified
+    //    Base priority = UN
+    //    Escalated priority = UN
+    // Case 2: Priority specified
+    //    Base priority = user specified priority
+    //    Escalated priority = UN
+    //
+    // Task will be created with max priority = max(base priority, UN) = base
+    // priority. We shouldn't need to do any additional manipulations here since
+    // basePriority should already be the right value
+    SWIFT_TASK_DEBUG_LOG("Creating detached task from current_task %p.", currentTask);
+    SWIFT_TASK_DEBUG_LOG("New detached task base pri = %zu", basePriority);
+
+  } else if (swift_task_is_unstructured(jobFlags)) {
```

Comment:
The two clauses for non-detached tasks are identical in semantics except for logging messages and the assertion that structured child tasks always have a parent.

Reply:
Yes agreed. I think I still prefer it this way since the semantics of what the escalated priority in the ActiveTaskStatus will be for them is different - structured concurrency tasks gets the escalated priority from the parent when the status record gets linked into the parent. Unstructured tasks don't and have escalated priority set to base priority. 

I've documented it as such in the comments.